sdk/core/src/main/java/com/klaviyo/coresdk/networking/KlaviyoApiClient.kt

```diff
@@ -0,0 +1,180 @@
+        Registry.dataStore.fetch(QUEUE_KEY)?.let {
+            val queue = JSONArray(it)
+            Array(queue.length()) { i -> queue.optString(i) }
+        }?.map { uuid ->
+            Registry.dataStore.fetch(uuid)?.let { json ->
+                val jsonObject = JSONObject(json)
+                val request = KlaviyoApiRequest.fromJson(jsonObject)
+                enqueueRequest(request)
+            }
+        }
+    }
+
+    /**
+     * Initialize the network queue batching with a thread to run on
+     */
+    private fun initBatch() {
+        if (!handlerThread.isAlive) {
+            handlerThread.start()
+            handler = HandlerUtil.getHandler(handlerThread.looper)
+        }
+
+        handler?.post(NetworkRunnable())
+    }
+
+    /**
+     * Gets the current size of the API queue
+     *
+     * @return number of requests in the batch queue
+     */
+    fun getQueueSize(): Int = apiQueue.size
+
+    /**
+     * Force the API queue to send immediately instead of waiting on the configured criteria
+     */
+    private fun flushQueue() {
+        handler?.removeCallbacksAndMessages(null)
+        handler?.post(NetworkRunnable(true))
+    }
+
+    /**
+     * Runnable which flushes the API queue in batches for efficiency.
+     * As long as there are items in the queue, the thread will loop.
+     * When the queue is cleared the thread will not loop itself and will terminate.
+     *
+     * @property force Boolean that will force the queue to flush now
+     */
+    class NetworkRunnable(val force: Boolean = false) : Runnable {
+        private val queueInitTime = Registry.clock.currentTimeMillis()
+
+        private val flushInterval: Long
+            get() = Registry.config.networkFlushInterval.toLong()
+
+        private val flushDepth: Int
+            get() = Registry.config.networkFlushDepth
+
+        override fun run() {
+            val emptied = flushQueue(force)
+            if (!emptied) {
+                handler?.postDelayed(this, flushInterval)
+            }
+        }
+
+        /**
+         * Empties all queued requests by instantly sending each request to Klaviyo
+         * The queue will flush whenever the triggers specified in config are met
+         *
+         * @param force Overrides the typical timings to force the queue to flush
+         *
+         * @return Whether the request queue was emptied or not
+         */
+        private fun flushQueue(force: Boolean = false): Boolean {
+            val queueTimePassed = Registry.clock.currentTimeMillis() - queueInitTime
+
+            if (force || getQueueSize() >= flushDepth || queueTimePassed >= flushInterval) {
+                while (!apiQueue.isEmpty()) {
+                    val apiRequest = apiQueue.poll()
+                    apiRequest?.apply {
+                        send()
+                        Registry.dataStore.clear(uuid)
+                    }
+                }
+
+                Registry.dataStore.clear(QUEUE_KEY)
```

Comment:
What thread is this work performed on ? Can something be added to the queue after you exit the while loop but before this is called.

Reply:
As I understand it, everything in `NetworkRunnable` will be running in our BG thread. Concurrency of the persisted API queue is definitely a concern worth investigating though

Reply:
If I just write the content of the concurrent queue here, instead of simply clearing, that should be safe...

Reply:
#33